File: Code/Avernakis Nodejs/Avernakis-Nodejs/Private/UiCalendar.cpp
#include "StdAfx.h"
#include "UiCalendar.h"

#define ThisMethod($x) &UiCalendar::$x
#define AutoAddMethod($x, ...) AddMethod<__VA_ARGS__>( #$x, ThisMethod( $x ) )
#define MakeThisFunc($x) MakeFunc( this, &UiCalendar::$x )

namespace Nav
{

	namespace
	{
		ObjectRegister<UiCalendar> c_obj;
	}

	void UiCalendar::DefineControl()
	{
		AutoAddMethod( SetRange );
		AutoAddMethod( GetRange, WrapObjectGeneric );

		AutoAddMethod( SetRangeLevel );
		AutoAddMethod( GetRangeLevel, WrapObjectGeneric );

		AutoAddMethod( SetDate );
		AutoAddMethod( GetDate, WrapObjectGeneric );

		AutoAddMethod( SetDateMark );
		AutoAddMethod( GetDateMark, WrapObjectGeneric );

		AutoAddMethod( SetCultureId );
		AutoAddMethod( GetCultureId, WrapObjectGeneric );

		AutoAddMethod( SetFormat );
		AutoAddMethod( GetFormat, WrapObjectGeneric );

		AutoAddMethod( OnChange, WrapObjectGeneric );
	}

	U1 UiCalendar::Ctor( UiWindow * p, Napi::Value v )
	{
		if ( !__CreateControl( p, v ) )
			return false;

		return true;
	}

	void UiCalendar::__OnChange( Ui::ICalendar & sender )
	{
		m_OnChange( this );
	}

	UiCalendar * UiCalendar::SetFormat( const WrapData<DateTimeFormat>& pFormat )
	{
		if ( 7 != pFormat.m_DayNameFull.Size() || 7 != pFormat.m_DayNameAbbreviated.Size() || 12 != pFormat.m_MonthNameFull.Size() || 12 != pFormat.m_MonthNameAbbreviated.Size() )
			return this;

		m_Format = pFormat;

		CultureInfoDateTime cidt{};
		m_Format.ToCultureInfo( cidt );
		GetControlTyped().SetFormat( cidt );

		return this;
	}

	WrapData<DateTimeFormat> UiCalendar::GetFormat() const
	{
		WrapData<DateTimeFormat> r{};
		r.FromCultureInfo( GetControlTyped().GetFormat() );
		return r;
	}

	UiCalendar * UiCalendar::OnChange( OnChange_t && fn )
	{
		SetEventCallback<Ui::ICalendar::OnChange>( m_OnChange, std::move( fn ), MakeThisFunc( __OnChange ) );
		return this;
	}

}

File: Code/Avernakis Nodejs/Test-Nodejs/src/Ave/Ui/Control/UiCalendar.ts
import { IControl } from "../UiControl";
import { AveLib } from "../../AveLib";
import { WindowLike } from "./UiWindow";
import { TimePoint } from "../../TimePoint";
import { CultureId, CultureInfoDateTime } from "../../Culture";

export enum CalendarRangeLevel {
	Month,
	Year,
	Decade,
	Century,
	Millenary,
}

export interface ICalendar extends IControl {
	new (window: WindowLike): ICalendar;

	SetRange(tpFrom: TimePoint, tpTo: TimePoint): Calendar;
	GetRange(): TimePoint[];

	SetRangeLevel(n: CalendarRangeLevel): Calendar;
	GetRangeLevel(): CalendarRangeLevel;

	SetDate(n: TimePoint): Calendar;
	GetDate(): TimePoint;

	SetDateMark(c: TimePoint): Calendar;
	GetDateMark(): TimePoint;

	SetCultureId(n: CultureId): Calendar;
	GetCultureId(): CultureId;

	SetFormat(n: CultureInfoDateTime): Calendar;
	GetFormat(): CultureInfoDateTime;

	OnChange(fn: (sender: Calendar) => void): Calendar;
}

export class Calendar extends (AveLib.UiCalendar as ICalendar) {
	GetDate() {
		return TimePoint.FromTick(super.GetDate().Tick);
	}

	GetDateMark() {
		return TimePoint.FromTick(super.GetDateMark().Tick);
	}
}
